### PR TITLE
Perform CliffordState measurements via method dispatch

### PR DESCRIPTION
**Pull request by smitsanghavi**
And deprecate the old `perform_measurement` method that called the `_measure` method directly.

Also deprecate all uses of `CliffordTableau` inside the `CliffordState`. Full clean up will follow.

#2948 #2423

### CODE REVIEW

File: cirq/sim/clifford/clifford_simulator.py [hunk 1]
```diff
@@ -354,6 +372,7 @@ def apply_unitary(self, op: 'cirq.Operation'):
         rewrite=lambda args, kwargs: (args, {('collapse_state_vector' if k ==
                                               'collapse_wavefunction' else k): v
                                              for k, v in kwargs.items()}))
+    @deprecated(deadline='v0.10.0', fix='Use the apply_measurement instead')
```
Comment (smitsanghavi) [inline cirq/sim/clifford/clifford_simulator.py:375]: Is the `CliffordState` directly used by end users? Do they call the `perform_measurement` function? If not, I can just delete it.

File: cirq/sim/clifford/clifford_simulator.py [hunk 2]
```diff
@@ -346,6 +344,26 @@ def apply_unitary(self, op: 'cirq.Operation'):
...
+        qids = [self.qubit_map[i] for i in op.qubits]
+
+        args = clifford.ActOnCliffordTableauArgs(state.tableau, qids, prng,
+                                                 measurements)
+        act_on(op, args)
+
+        for i, qid in enumerate(qids):
+            state.ch_form.project_Z(qid, measurements[key][i])
+
     @deprecated_parameter(
```
Comment (smitsanghavi) [inline cirq/sim/clifford/clifford_simulator.py:367]: Given that the whole function is deprecated now, should I remove the argument deprecation notice?

**Comment by smitsanghavi** [comment]
Friendly Ping!

**Comment by smitsanghavi** [comment]
Holding off on this PR for now

File: cirq/sim/clifford/clifford_simulator.py [hunk 3]
```diff
@@ -312,11 +312,13 @@ def __str__(self) -> str:
     def to_numpy(self) -> np.ndarray:
         return self.ch_form.to_state_vector()
 
+    @deprecated(deadline='v0.10.0', fix='use CliffordTableau instead')
```
Comment (balopat) [inline cirq/sim/clifford/clifford_simulator.py:315]: This should be deprecated in 0.11.0 at least - we are in 0.10.0.dev on master!

File: cirq/sim/clifford/clifford_simulator.py [hunk 4]
```diff
@@ -312,11 +312,13 @@ def __str__(self) -> str:
...
     def stabilizers(self) -> List[DensePauliString]:
         """Returns the stabilizer generators of the state. These
         are n operators {S_1,S_2,...,S_n} such that S_i |psi> = |psi>"""
         return self.tableau.stabilizers()
 
+    @deprecated(deadline='v0.10.0', fix='use CliffordTableau instead')
```
Comment (balopat) [inline cirq/sim/clifford/clifford_simulator.py:321]:
```suggestion
    @deprecated(deadline='v0.11.0', fix='use CliffordTableau instead')
```

File: cirq/sim/clifford/clifford_simulator.py [hunk 5]
```diff
@@ -359,6 +374,7 @@ def apply_unitary(self, op: 'cirq.Operation'):
             },
         ),
     )
+    @deprecated(deadline='v0.10.0', fix='Use the apply_measurement instead')
```
Comment (balopat) [inline cirq/sim/clifford/clifford_simulator.py:377]:
```suggestion
    @deprecated(deadline='v0.11.0', fix='Use the apply_measurement instead')
```

File: cirq/sim/clifford/clifford_simulator.py (hunk 1 above)
Comment (balopat) [reply cirq/sim/clifford/clifford_simulator.py:377]: Let's just be on the safe side, it's a public function. We don't have usage numbers!

**Comment by smitsanghavi** [comment]
Ping!

File: cirq/sim/clifford/clifford_simulator.py [hunk 6]
```diff
@@ -331,21 +333,34 @@ def wave_function(self):
...
         return
 
+    def apply_measurement(
+        self,
+        op: 'cirq.Operation',
+        measurements: Dict[str, List[np.ndarray]],
+        prng: np.random.RandomState,
+        collapse_state_vector=True,
+    ):
+        if collapse_state_vector:
```
Comment (balopat) [inline cirq/sim/clifford/clifford_simulator.py:354]: as this is a public method, I think we should ensure that the op passed in has a MeasurementGate as op.gate. WDYT?

Comment (smitsanghavi) [reply cirq/sim/clifford/clifford_simulator.py:360]: Makes sense. Added the validation and some tests.

**Review by balopat — APPROVED** [review]
LGTM, thank you!

**Pull request merged**